modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/AddWriteInvokeClosure.java

```diff
@@ -0,0 +1,154 @@
+ *
+ * <p>Synchronization between reading and updating the version chain occurs due to the locks (read and write) of the page of the tree on
+ * which the version chain is located.
+ *
+ * <p>Synchronization between update operations for the version chain must be external (by {@link RowId row ID}).
+ *
+ * <p>Operation may throw {@link StorageException} and {@link TxIdMismatchException} which will cause form
+ * {@link BplusTree#invoke(Object, Object, InvokeClosure)}.
+ */
+class AddWriteInvokeClosure implements InvokeClosure<VersionChain> {
+    private final RowId rowId;
+
+    private final @Nullable BinaryRow row;
+
+    private final UUID txId;
+
+    private final UUID commitTableId;
+
+    private final int commitPartitionId;
+
+    private final AbstractPageMemoryMvPartitionStorage storage;
+
+    private @Nullable VersionChain newRow;
+
+    private @Nullable BinaryRow previousUncommittedRowVersion;
+
+    private @Nullable RowVersion toRemove;
+
+    AddWriteInvokeClosure(
+            RowId rowId,
+            @Nullable BinaryRow row,
+            UUID txId,
+            UUID commitTableId,
+            int commitPartitionId,
+            AbstractPageMemoryMvPartitionStorage storage
+    ) {
+        this.rowId = rowId;
+        this.row = row;
+        this.txId = txId;
+        this.commitTableId = commitTableId;
+        this.commitPartitionId = commitPartitionId;
+        this.storage = storage;
+    }
+
+    @Override
+    public void call(@Nullable VersionChain oldRow) throws IgniteInternalCheckedException {
+        if (oldRow == null) {
+            RowVersion newVersion = insertRowVersion(row, NULL_LINK);
+
+            newRow = VersionChain.createUncommitted(rowId, txId, commitTableId, commitPartitionId, newVersion.link(), NULL_LINK);
+
+            return;
+        }
+
+        if (oldRow.isUncommitted()) {
+            throwIfChainBelongsToAnotherTx(oldRow, txId);
+        }
+
+        RowVersion newVersion = insertRowVersion(row, oldRow.newestCommittedLink());
+
+        if (oldRow.isUncommitted()) {
+            RowVersion currentVersion = storage.readRowVersion(oldRow.headLink(), ALWAYS_LOAD_VALUE);
+
+            previousUncommittedRowVersion = storage.rowVersionToBinaryRow(currentVersion);
+
+            // As we replace an uncommitted version with new one, we need to remove old uncommitted version.
+            toRemove = currentVersion;
+        }
+
+        newRow = VersionChain.createUncommitted(rowId, txId, commitTableId, commitPartitionId, newVersion.link(), newVersion.nextLink());
+    }
+
+    @Override
+    public @Nullable VersionChain newRow() {
+        assert newRow != null;
+
+        return newRow;
+    }
+
+    @Override
+    public OperationType operationType() {
+        return OperationType.PUT;
+    }
+
+    @Override
+    public void onUpdate() {
+        if (toRemove != null) {
+            storage.removeRowVersion(toRemove);
```

Comment:
Again, I recommend removing the data when you don't hold any locks.
Maybe you can even read the value when holding no locks, it should be safe.

Reply:
I agree, I'll fix it.